### ISSUE TITLE
Low performance on ML-1m dataset

### ISSUE THREAD

**Issue by abstrcode** [RUCAIBox/RecBole#945]
Hello,

I simply run the GRU4Rec algorithm on the ML-1m dataset by modifying the provided script example. However, the performance is rather low compared to the GRU4Rec result in BERT4Rec paper.  The runned final test result is : {'recall@10': 0.2548, 'mrr@10': 0.1059, 'ndcg@10': 0.1406, 'hit@10': 0.2548, 'precision@10': 0.0255}. I wonder what is your test result for GRU4Rec on ML-1m or the possible reasons. Thanks.

**Comment by hyp1231** [comment]
Hi, could you please provide your `.yaml` file?

I guess there is inconsistency of dataset preparation and model hyper-parameters between the test and the original paper.

**Comment by abstrcode** [comment]
@hyp1231 the .yaml file is
`# general
#gpu_id: 0,1,2,3,4,5,6,7
gpu_id: 0
use_gpu: True
seed: 2020
state: INFO
reproducibility: True
data_path: 'path to ML-1m data'
checkpoint_dir: 'saved'
show_progress: True

# training settings
epochs: 300
train_batch_size: 2048
learner: adam
learning_rate: 0.001
training_neg_sample_num: 0
training_neg_sample_distribution: uniform
eval_step: 1
stopping_step: 10
clip_grad_norm: ~
# clip_grad_norm:  {'max_norm': 5, 'norm_type': 2}
weight_decay: 0.0
draw_loss_pic: False

# evaluation settings
train_split: 1
eval_setting: TO_LS,full
group_by_user: True
split_ratio: [0.8,0.1,0.1]
leave_one_num: 2
real_time_process: False
metrics: ["Recall","MRR","NDCG","Hit","Precision"]
topk: [10]
valid_metric: MRR@10
valid_metric_bigger: True
eval_batch_size: 1
loss_decimal_place: 4
metric_decimal_place: 4`

**Comment by hyp1231** [comment]
Hi, thanks for the detailed config file.

1. In BERT4Rec paper, "we keep users with at least five feedbacks", thus we need to set `min_user_inter_num: 5` in RecBole.
2. In BERT4Rec paper, "pairing each ground truth item in the test set with 100 randomly sampled negative items that the user has not interacted with", while in the `.yaml` provided above, the evaluation follows a full sort prediction, and we need to set `eval_setting: TO_LS,pop100`.

There may be other inconsistency as I have not carefully check all the parameters. I suggest detailed and thoroughly check of dataset preparation, evaluation setting, statistics of processed datasets, etc.

**Comment by abstrcode** [comment]
@hyp1231 Thanks for the suggestion. I tried the new new dataset setting and the eval_setting. However, the result is even worse and far bellow the reported result. New test result: {'recall@10': 0.0997, 'mrr@10': 0.0294, 'ndcg@10': 0.0455, 'hit@10': 0.0997, 'precision@10': 0.01}.  It seems other issue reported similar worse result as [Issues 911](https://github.com/RUCAIBox/RecBole/issues/911).

**Comment by hyp1231** [comment]
> There may be **other inconsistency** since I have not carefully check all the parameters. I suggest **detailed and thorough** check of dataset preparation, evaluation setting, statistics of processed datasets, etc.

Directly involve these two modification is **not** enough. Please check all the details described in the paper you would like to reproduce.

The main issue lies in the truncated length of item sequences. In BERT4Rec paper Sec. 4.3:
> N = 200 for ML-1m and ML-20m

, while RecBole takes a default value of 50 if you don't specific it.

I test GRU4Rec on ML-1M using the lasted version of **branch `data`**, which will be released as a new version soon. And I get `{'ndcg@1': 0.2868, 'ndcg@5': 0.433, 'ndcg@10': 0.4665, 'hit@1': 0.2868, 'hit@5': 0.5639, 'hit@10': 0.6669}`, which are within the same order of magnitude as the results in BERT4Rec paper. The `.yaml` file I use is:
```yaml
neg_sampling: ~
load_col:
  inter: '*'
user_inter_num_interval: '[5,inf)'
eval_args:
  split:
    LS: valid_and_test
  group_by: user
  order: TO
  mode: pop100
metrics: [MRR , NDCG , Hit]
topk: [1, 5, 10]
MAX_ITEM_LIST_LENGTH: 200
```

Besides, due to details of data preprocessing, the test data cannot be exactly the same as those used in paper.

**Comment by hyp1231** [comment]
Closing due to inactivity. Please comment if you're still having issues.
